QR video scanning

sisou opened the pull request:
Fixes #364

This PR extends the `QrScanner` library with the ability to scan live video for QR codes, converted from https://github.com/nimiq/qr-scanner.

It then adds a `QrVideoScanner` component to encapsulate the scanner. Also includes already-translated strings, taken from the nimiq/vue-components translations.

~~There is a workaround for the "BarcodeScanner", which is sadly not yet recognized by Typescript and thus cannot be used for types.~~

sisou committed:
Upgrade QrScanner with video scanning capabilities

Co-authored-by: Daniel <github.danimoh@neverbox.com>

sisou committed:
Fix dependency checker for BarcodeDetector

sisou committed:
Enable performance optimization

sisou committed:
Add QrVideoScanner component

Co-authored-by: Daniel <github.danimoh@neverbox.com>

sisou committed:
Add camera access error handling

Co-authored-by: Daniel <github.danimoh@neverbox.com>

mraveux commented on src/lib/QrScanner.js:
```diff
@@ -1,141 +1,565 @@
+    /**
+     * @param {HTMLImageElement | HTMLVideoElement | string} imageOrFileOrUrl
+     * @param {RectArea} [scanRegion]
+     * @param {Promise<Worker>} [givenEngine] // TODO: Or BarcodeDetector
```
Is there really no way to use `BarcodeDetector` in JsDoc?

sisou replied:
It's not included in TS. And probably won't be until there's wider support: https://caniuse.com/?search=barcodedetector

sisou replied:
Daniel added a type definition, so now it's included in the types.

mraveux commented on src/lib/QrScanner.js:
```diff
@@ -1,141 +1,565 @@
+                constraintsToTry.forEach(constraint => {
+                    constraint.facingMode = facingMode;
                 });
             }
```
Couldn't this be simplified to the following?
```
constraintsToTry.forEach(constraint => {
    constraint.facingMode = exact ? { exact: facingMode } : facingMode;
})
```

mraveux committed:
QrScanner: minor cleanup during review

mraveux commented on src/lib/QrScanner.js:
```diff
@@ -1,141 +1,564 @@
+        return navigator.mediaDevices.getUserMedia({
+            video: constraintsToTry.shift(),
+        }).catch(() => this._getMatchingCameraStream(constraintsToTry));
+    }
```
Wouldn't it be cleaner as an async function?
```
async _getMatchingCameraStream(constraintsToTry) {
    if (!navigator.mediaDevices || constraintsToTry.length === 0) {
        return Promise.reject('Camera not found.');
    }
    try {
        return navigator.mediaDevices.getUserMedia({
            video: constraintsToTry.shift(),
        });
    } catch (e) {
        return await this._getMatchingCameraStream(constraintsToTry);
    }
}
```

sisou replied:
With a slight change, that the `navigator.mediaDevices.getUserMedia()` function must be `await`ed, but yes, that works.

mraveux reviewed (commented):
Looks nice, as always 👍 
The only thing that prevent me from approving the PR is that, as discussed, it doesn't seem to work as intended on Chrome android (and maybe other browsers but that'll require more testing) but that should not be a big issue to fix imo.

sisou committed:
Fix video scan region calculation

sisou committed:
Address code review suggestions

danimoh committed:
QrScanner: remove code not used in the Keyguard

mraveux reviewed (approved):
Perfect! It now works perfectly on mobile devices as well 😃

danimoh committed:
QrScanner: type improvements

danimoh committed:
QrScanner: performance improvements

- re-use a created qr engine and canvas in nested calls
- round scan region to whole pixel positions to avoid blurry
  drawing at sub-pixel positions
- avoid clearing the canvas by resetting the canvas width or
  height when they actually didn't change.

danimoh committed:
QrScanner: refactor scanImage with async/await

danimoh committed:
ImportFile: only scan valid login files via video qr scanner

danimoh committed:
ImportFile: add a tooltip to qr scanning button

danimoh reviewed (approved):
Nice work :+1:
I applied a variety of improvements and a fix for scanning invalid LoginFiles / arbitrary QR codes, see commit history.
Please have a look at the changes.
I also rebased the branch onto master to avoid later merge conflicts as I changed code touching the current LoginFile validation that already supports the encoded wallet label.
As the QrScanner and QrVideoScaner are largely adapted from the `qr-scanner` lib and `@nimiq/vue-components`, I took the freedom, to also add me as a co-author to the commits adding these chunks.
However, I'm also not upset if you object to this and want to force-push it away.

sisou committed:
Simplify types & constructor signature
